src/v/datalake/translation/partition_translator.cc

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "datalake/translation/partition_translator.h"
+
+#include "cluster/partition.h"
+#include "datalake/batching_parquet_writer.h"
+#include "datalake/coordinator/frontend.h"
+#include "datalake/coordinator/translated_offset_range.h"
+#include "datalake/data_writer_interface.h"
+#include "datalake/logger.h"
+#include "datalake/record_multiplexer.h"
+#include "datalake/translation/state_machine.h"
+#include "kafka/server/partition_proxy.h"
+#include "kafka/utils/txn_reader.h"
+#include "resource_mgmt/io_priority.h"
+#include "ssx/future-util.h"
+
+namespace datalake::translation {
+
+static constexpr std::chrono::milliseconds translation_jitter{500};
+static constexpr int max_attempts = 5;
+constexpr ::model::timeout_clock::duration wait_timeout = 5s;
+constexpr size_t max_rows_per_row_group = std::numeric_limits<size_t>::max();
+constexpr size_t max_bytes_per_row_group = std::numeric_limits<size_t>::max();
+
+partition_translator::partition_translator(
+  ss::lw_shared_ptr<cluster::partition> partition,
+  ss::sharded<coordinator::frontend>* frontend,
+  ss::sharded<features::feature_table>* features,
+  std::chrono::milliseconds translation_interval,
+  ss::scheduling_group sg,
+  size_t reader_max_bytes,
+  ssx::semaphore& parallel_translations)
+  : _term(partition->raft()->term())
+  , _partition(std::move(partition))
+  , _frontend(frontend)
+  , _features(features)
+  , _jitter{translation_interval, translation_jitter}
+  , _max_bytes_per_reader(reader_max_bytes)
+  , _parallel_translations(parallel_translations)
+  , _writer_scratch_space(std::filesystem::temp_directory_path())
+  , _logger(prefix_logger{
+      datalake_log, fmt::format("{}-term-{}", _partition->ntp(), _term)}) {
+    _stm = _partition->raft()
+             ->stm_manager()
+             ->get<datalake::translation::translation_stm>();
+    vassert(_stm, "No translation stm found for {}", _partition->ntp());
+    ssx::repeat_until_gate_closed(_gate, [this, sg] {
+        return ss::with_scheduling_group(sg, [this] { return do_translate(); });
+    });
```

Comment:
consider separating this into a `start()` method to be able to decouple creation from starting